proposals/0002-root-signature-in-clang.md

```diff
@@ -471,116 +489,173 @@ The additional semantic rules not already covered by the grammar are listed here
 ```
 
+The function/root signature associates a function (the first operand) with a
+reference to a root signature (the second operand).
+
+#### Root Signature
+
+```LLVM
+!3 = !{!4, !5, !6, !7 }
+```
+
+The root signature itself consists of a list of references to root signature
+elements.
+
+#### Root Signature Elements
+
+Root signature elements are identified by the first operand, which is a string.
+The following root signature elements are defined:
+
+* Root flags ("RootFlags")
+* Root constants ("RootConstants")
+* Root descriptors ("RootCBV", "RootSRV", "RootUAV")
+* Descriptor tables ("DescriptorTable")
+* Static samplers ("StaticSampler")
+
+As in the [string representation](#root-signature-grammar) of the root
+signature, the elements can appear in any order. This does mean that invalid
+root signatures can be represented in the metadata (eg multiple root flags). As
+a result of this anything that takes the metadata as input must validate the
+incoming metadata, even if the HLSL frontend happens to perform validation
+before generating the metadata.
+
+#### Root Flags
+
+```LLVM
+!4 = { !"RootFlags", i32 1 }
+```
+
+Operands:
+
+* i32: the root signature flags ([D3D12_ROOT_SIGNATURE_FLAGS](d3d12_root_signature_flags))
+
+
+#### Root Constants
+
+```LLVM
+!123 = { !"RootConstants", i32 0, i32 1, i32 2, i32 3 }
+```
+
+Operands:
+* i32: shader visibility ([D3D12_SHADER_VISIBILITY][d3d12_shader_visibility])
+* i32: shader register
+* i32: register space
+* i32: number 32 bit values
+
+#### Root descriptors
+
+Root descriptors come in three flavors, but they have the same structure. The
+flavors are:
+
+* Root constant buffer view ("RootCBV")
+* Root shader resource view ("RootSRV")
+* Root unordered access view ("RootUAV")
+
+```LLVM
+!5 = { !"RootCBV", i32 0, i32 1, i32 0, i32 0 }
+```
+
+Operands:
+* i32: shader visibility  ([D3D12_SHADER_VISIBILITY][d3d12_shader_visibility])
+* i32: shader register
+* i32: register space
+* i32: root descriptor flags ([D3D12_ROOT_DESCRIPTOR_FLAGS][d3d12_root_descriptor_flags])
+
+#### Descriptor Tables
+
+Descriptor tables are made up of descriptor ranges.
+
+```LLVM
+!7 = { !"DescriptorTable", i32 0, !8, !9 }
+```
+
+Operands:
+
+* i32: shader visibility  ([D3D12_SHADER_VISIBILITY][d3d12_shader_visibility])
+* remaining operands are references to descriptor ranges
+
+##### Descriptor Ranges
+
+```LLVM
+!8 = !{ !"SRV", i32 0, i32 0, i32 -1, i32 0 }
+!9 = !{ !"UAV", i32 5, i32 1, i32 10, i32 0 }
+```
+
+Operands:
+
+* string: type of range - "SRV", "UAV", "CBV" or "Sampler"
+* i32: number of descriptors in the range
+* i32: base shader register
+* i32: register space
+* i32: descriptor range flgs ([D3D12_DESCRIPTOR_RANGE_FLAGS][d3d12_descriptor_range_flags])
+
+#### Static Samplers
+
+```LLVM
+!6 = !{ !"StaticSampler", i32 1, i32 0, ... }; remaining operands omitted for space
+```
+
+Operands:
+* i32: Filter ([D3D12_FILTER](d3d12_filter))
```

Comment:
These also use regular braces as opposed to square. Not sure which one is correct but I'll note it to double-check.